File: python/level2/exhaustiveSearch_42842.py
# https://programmers.co.kr/learn/courses/30/lessons/42842

# 브라운 : 2x + 2y - 4
# 노랑 : (x-2)*(y-2)
# 전체 = x*y = 가로*세로 => 따라서 전체는 가로로 나누어진다고 볼 수 있음
def solution(brown, yellow):
    total = brown + yellow
    for width in range(total, 2, -1): # 가로
        if total % width == 0:
            height = total // width # 세로를 탐색하는거라서 굳이 나머지가 0인지는 따지지않고 몫만 가져와서 아래에서 따져봄
            if yellow == (width - 2) * (height - 2):
                return [width, height]

# 근의 공식으로 해결
# x : 가로, y : 세로
# x*y = brown + yellow
# (x-2)*(y-2) = yellow
# def solution(brown, yellow):
#     x = (brown + 4 +((brown+4) ** 2- 16*(brown+red)**0.5)/4)
#     y = (brown + red) // width
#     return [max(x, y), min(x, y)]

brown = 24
yellow = 24
print(solution(brown, yellow))

File: java/src/solution/level2/exhaustiveSearch_42842.java
//https://programmers.co.kr/learn/courses/30/lessons/42842?language=java
package solution.level2;
import java.util.Arrays;

// 규칙
// x : 가로, y: 세로
// brown 갯수 = 2x + 2y - 4  <- -4는 꼭지점 가로세로 2x, 2y로 갯수 구하면 겹치는 부분 @@@ 이 공식 구하는게 핵심@@@
    // 위 식에서 x = (brown갯수 + 4) / 2 - y ---> line 17
    // 위 식에서 y 값은 (brown + 4) / 2 보다는 무조건 작아야함을 알 수 있다 ---> line 16에 사용
// yellow 갯수 = (x-2) * (y-2)
// 카펫 세로길이 <= 가로길이 ---> 이부분을 응용해서 line 1
// height 최소값 = 3
public class exhaustiveSearch_42842 {
    public static int[] solution(int brown, int yellow) {
        int height = 0;
        int width = 0;
        for (height = 3; height < (int) (brown + 4) / 2; height++) {

            width = ((brown + 4) / 2) - height;
            if (width < height) { // 이조건에 걸리면 for문 반복하는 의미가 없음
                System.out.println("height greater than width. Impossible condition");
                break;
            }

            int yellowCount = (width - 2) * (height - 2);
            if (yellow == yellowCount) { // 이조건이 안맞으면 for 문 다시 돌아야함
                break;
            }
        }
        int[] answer = new int[] { width, height };
        return answer;
    }
    public static void main(String[] args){
        int brown = 24;
        int yellow = 24;
        int[] answer = solution(brown, yellow);
        System.out.println(Arrays.toString(answer));
    }
}
